ai/smart-wallet-example/smart-wallet-core-solidty/contracts/OTKAccount.sol

```diff
@@ -0,0 +1,238 @@
+    /// @param _transaction The transaction to execute.
+    function executeTransactionFromOutside(Transaction calldata _transaction) external payable override {
+        // Behave the same as for fallback/receive, just execute nothing, returns nothing
+    }
+
+    /// @notice Inner method for executing a transaction.
+    /// @param _transaction The transaction to execute.
+    function _execute(Transaction calldata _transaction) internal {
+        address to = address(uint160(_transaction.to));
+        uint128 value = Utils.safeCastToU128(_transaction.value);
+        bytes calldata data = _transaction.data;
+        uint32 gas = Utils.safeCastToU32(gasleft());
+
+        // Note, that the deployment method from the deployer contract can only be called with a "systemCall" flag.
+        bool isSystemCall;
+        if (to == address(DEPLOYER_SYSTEM_CONTRACT) && data.length >= 4) {
+            bytes4 selector = bytes4(data[:4]);
+            // Check that called function is the deployment method,
+            // the others deployer method is not supposed to be called from the default account.
+            isSystemCall =
+                selector == DEPLOYER_SYSTEM_CONTRACT.create.selector ||
+                selector == DEPLOYER_SYSTEM_CONTRACT.create2.selector ||
+                selector == DEPLOYER_SYSTEM_CONTRACT.createAccount.selector ||
+                selector == DEPLOYER_SYSTEM_CONTRACT.create2Account.selector;
+        }
+        bool success = EfficientCall.rawCall(gas, to, value, data, isSystemCall);
+        if (!success) {
+            EfficientCall.propagateRevert();
+        }
+    }
+
+    /// @notice Validation that the ECDSA signature of the transaction is correct.
+    /// @param _hash The hash of the transaction to be signed.
+    /// @param _signature The signature of the transaction.
+    /// @return EIP1271_SUCCESS_RETURN_VALUE if the signaure is correct. It reverts otherwise.
+    function _isValidSignature(bytes32 _hash, bytes memory _signature) internal returns (bool) {
+        require(_signature.length == 65, "Signature length is incorrect");
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        // Signature loading code
+        // we jump 32 (0x20) as the first slot of bytes contains the length
+        // we jump 65 (0x41) per signature
+        // for v we load 32 bytes ending with v (the first 31 come from s) then apply a mask
+        assembly {
+            r := mload(add(_signature, 0x20))
+            s := mload(add(_signature, 0x40))
+            v := and(mload(add(_signature, 0x41)), 0xff)
+        }
+        require(v == 27 || v == 28, "v is neither 27 nor 28");
+
+        address recoveredAddress = ecrecover(_hash, v, r, s);
+
+        if (recoveredAddress == address(this) && recoveredAddress != address(0)) {
+            return true;
+        }
+
+        if (otks[recoveredAddress] == true) {
+            otks[recoveredAddress] = false;
+            return true;
+        }
```

Comment:
This one-time key validation is vulnerable to front-running attacks. An attacker could observe a valid signature transaction in the mempool and front-run it with a higher gas price.

Reply:
@thomas-nguy

Reply:
@thomas-nguy @leejw51crypto any more comments on this one?

Reply:
nope, can be done as another task